### PR TITLE
Add endpoint disambiguation

### PR DESCRIPTION
**Pull request by rynowak**
- Better sample of metadata
- Sample shows how conventional routing would work
- Added endpoint disambiguation

### CODE REVIEW

File: samples/DispatcherSample/HttpMethodEndpointSelector.cs [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Threading.Tasks;
+using Microsoft.AspNetCore.Dispatcher;
+
+namespace DispatcherSample
+{
+    public class HttpMethodEndpointSelector : EndpointSelector
```
Comment (rynowak) [inline samples/DispatcherSample/HttpMethodEndpointSelector.cs]: This is an off the cuff implementation from memory of MVC's http method selection + backtracking

File: src/Microsoft.AspNetCore.Dispatcher/DispatcherValueEndpointSelector.cs [hunk 2]
```diff
@@ -0,0 +1,55 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Threading.Tasks;
+
+namespace Microsoft.AspNetCore.Dispatcher
+{
+    public class DispatcherValueEndpointSelector : EndpointSelector
```
Comment (rynowak) [inline src/Microsoft.AspNetCore.Dispatcher/DispatcherValueEndpointSelector.cs]: This is a brute force (and wrong) implementation. I wanted to save the real thing for @jbagga

File: src/Microsoft.AspNetCore.Routing/Dispatcher/RouteTemplateDispatcher.cs [hunk 3]
```diff
@@ -0,0 +1,157 @@
...
+using System;
+using System.Collections.Generic;
+using System.Threading.Tasks;
+using Microsoft.AspNetCore.Dispatcher;
+using Microsoft.AspNetCore.Http;
+using Microsoft.AspNetCore.Routing.Template;
+
+namespace Microsoft.AspNetCore.Routing.Dispatcher
+{
+    public class RouteTemplateDispatcher : DispatcherBase
```
Comment (rynowak) [inline src/Microsoft.AspNetCore.Routing/Dispatcher/RouteTemplateDispatcher.cs]: This is mostly a copy-paste of all the boilerplate from `RouteBase`

File: src/Microsoft.AspNetCore.Dispatcher.Abstractions/IDispatcherFeature.cs [hunk 4]
```diff
@@ -10,5 +10,7 @@ public interface IDispatcherFeature
         Endpoint Endpoint { get; set; }
 
         RequestDelegate RequestDelegate { get; set; }
+
+        DispatcherValueCollection Values { get; set; }
```
Comment (davidfowl) [inline src/Microsoft.AspNetCore.Dispatcher.Abstractions/IDispatcherFeature.cs]: This looks suspect

Comment (rynowak) [reply src/Microsoft.AspNetCore.Dispatcher.Abstractions/IDispatcherFeature.cs]: Let's discuss this. We need to make a call about whether or not route values are first class

**Pull request merged**